algebra-cl-gen/src/gpu/sources.rs

```diff
@@ -0,0 +1,436 @@
+use algebra::{PrimeField, AffineCurve, Fp2Parameters};
+use itertools::*;
+use ff_cl_gen as ffgen;
+use ffgen::*;
+
+use std::any::TypeId;
+use std::collections::HashMap;
+
+// Instead of having a very large OpenCL program written for a specific curve, with a lot of
+// rudandant codes (As OpenCL doesn't have generic types or templates), this module will dynamically
+// generate OpenCL codes given different PrimeFields and curves.
+
+static FFT_SRC: &str = include_str!("fft/fft.cl");
+static FIELD2_SRC: &str = include_str!("multiexp/field2.cl");
+static EC_SRC: &str = include_str!("multiexp/ec.cl");
+static MULTIEXP_SRC: &str = include_str!("multiexp/multiexp.cl");
+static POLYCOMMIT_ROUND_REDUCE_SRC: &str = include_str!("polycommit/round_reduce.cl");
+
+fn params2<P, L: Limb>(field2: &str, field: &str) -> String
+where
+    P: Fp2Parameters
+{
+    let nonresidue = P::NONRESIDUE;
+    let nonresidue_def = define_field("FIELD_NONRESIDUE", L::limbs_of(nonresidue));
+    join(
+        &[
+            nonresidue_def,
+        ],
+        "\n",
+    )
+        .replace("FIELD2", field2)
+        .replace("FIELD", field)
+}
+
+fn field2(field2: &str, field: &str) -> String {
+    String::from(FIELD2_SRC)
+        .replace("FIELD2", field2)
+        .replace("FIELD", field)
+}
+
+fn fft(field: &str) -> String {
+    String::from(FFT_SRC).replace("FIELD", field)
+}
+
+
+fn polycommit_round_reduce(point: &str, exp: &str) -> String {
+    String::from(POLYCOMMIT_ROUND_REDUCE_SRC)
+        .replace("EXPONENT", exp)
+        .replace("POINT", point)
+}
+
+
+#[cfg(not(feature = "blstrs"))]
+const BLSTRS_DEF: &str = "";
+#[cfg(feature = "blstrs")]
+const BLSTRS_DEF: &str = "#define BLSTRS";
```

Comment:
What is this ? Do we need it ? I don't find any feature "blstrs" in the Cargo.toml

Reply:
Yes, we don't need it, I've removed it from the code